File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Ruby/remove_duplicates_from_sorted_array_II.rb
# Follow up for "Remove Duplicates":
#   What if duplicates are allowed at most twice?
#
# For example,
#   Given sorted array nums = [1,1,1,2,2,3],
#
#   Your function should return length = 5, with the first five elements of nums being 1, 1, 2, 2 and 3. It doesn't matter what you leave beyond the new length.

# @param {Integer[]} nums
# @return {Integer}
def remove_duplicates(nums)
  if nums.length < 3
    return nums.length
  end

  prev = nums[0]
  twiced = false
  it = 1

  while it < nums.length
    if twiced && nums[it] == prev
      nums.delete_at(it)
    elsif nums[it] == prev
      twiced = true
      it = it + 1
    else
      prev = nums[it]
      twiced = false
      it = it + 1
    end
  end

  return nums.length
end

nums = [1,1,1,2,2,3,3,3,3,3,4,5,5,5,5]
puts nums
remove_duplicates(nums)
puts "==="
puts nums

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/RemoveDuplicatesFromSortedArrayII.py
"""
 Follow up for "Remove Duplicates":
 What if duplicates are allowed at most twice?

 For example,
 Given sorted array A = [1,1,1,2,2,3],

 Your function should return length = 5, and A is now [1,1,2,2,3]. 
"""

        

class Solution:
    # @param A a list of integers
    # @return an integer
    def removeDuplicates(self, A):
        show_twice = False
        if len(A) < 3:
            return len(A)
        prev = A[0]
        remove_list = list() 
        for idx in range(1,len(A)):
            if A[idx] == prev:
                if show_twice:
                    remove_list.append(idx) 
                    #print "pop" + str(idx)
                else:
                    show_twice = True
                    #print "show"
            else:
                prev = A[idx]
                show_twice = False
                #print "move"
        remove_list.reverse()
        for idx in remove_list:
            A.pop(idx)
        return len(A)

if __name__ == "__main__":
    A=[1,1,1,1,1,1]
    inst = Solution()
    print inst.removeDuplicates(A)
    print A
